gpytorch/mlls/variational_elbo.py

```diff
@@ -1,94 +1,93 @@
+
+        log_likelihood = self.likelihood.variational_log_probability(variational_dist_f, target, **kwargs).div(
+            num_batch
+        )
+
+        kl_divergence = torch.distributions.kl.kl_divergence(variational_dist_u, prior_dist)
+
+        if kl_divergence.dim() > log_likelihood.dim():
+            kl_divergence = kl_divergence.sum(-1)
+
+        if log_likelihood.numel() == 1:
+            kl_divergence = kl_divergence.sum()
+
+        kl_divergence = kl_divergence.div(self.num_data)
+
+        if self.combine_terms:
+            res = log_likelihood - kl_divergence
+            for _, prior, params_and_tfs, prior_tf in self.named_priors():
+                params = [p if ptf is None else ptf(p) for p, ptf in params_and_tfs]
+                lp_arg = params[0] if prior_tf is None else prior_tf(*params)
+                res.add_(prior.log_prob(lp_arg).sum().div(self.num_data))
+            return res
+        else:
+            log_prior = torch.zeros_like(log_likelihood)
+            for _, prior, params_and_tfs, prior_tf in self.named_priors():
+                params = [p if ptf is None else ptf(p) for p, ptf in params_and_tfs]
+                lp_arg = params[0] if prior_tf is None else prior_tf(*params)
+                log_prior.add_(prior.log_prob(lp_arg).sum())
+            return log_likelihood, kl_divergence, log_prior.div(self.num_data)
+
+
+class VariationalELBOEmpirical(VariationalELBO):
+    def __init__(self, likelihood, model, num_data):
+        """
+        A special MLL designed for variational inference.
+        This computes an empirical (rather than exact) estimate of the KL divergence
+
+        Args:
+        - likelihood: (Likelihood) - the likelihood for the model
+        - model: (Module) - the variational GP model
+        - num_data: (int) - the total number of training data points (necessary for SGD)
+        """
+        super(VariationalELBOEmpirical, self).__init__(likelihood, model, num_data, combine_terms=True)
+
+    def forward(self, variational_dist_f, target, **kwargs):
+        num_batch = variational_dist_f.event_shape[0]
+        variational_dist_u = self.model.variational_strategy.variational_distribution.variational_distribution
+        prior_dist = self.model.variational_strategy.prior_distribution
+
+        log_likelihood = self.likelihood.variational_log_probability(variational_dist_f, target, **kwargs)
+        log_likelihood = log_likelihood.div(num_batch)
+
+        num_samples = settings.num_likelihood_samples.value()
+        variational_samples = variational_dist_u.rsample(torch.Size([num_samples]))
+        kl_divergence = (
+            variational_dist_u.log_prob(variational_samples) - prior_dist.log_prob(variational_samples)
+        ).mean(0)
+        kl_divergence = kl_divergence.div(self.num_data)
+
+        res = log_likelihood - kl_divergence
+        for _, prior, params, tf in self.named_priors():
+            lp_arg = params[0] if tf is None else tf(*params)
```

Comment:
This bit of code is different than the one for VariationalELBO

Reply:
so this should implement the same functionality as before - are you saying there is something weird unrelated to this PR?